src/Advection/vector_invariant_advection.jl

```diff
@@ -38,142 +32,173 @@ Keyword arguments
-
-# Make sure that divergence is absorbed in the vertical scheme is 1. divergence_schem == Nothing 2. vertical_scheme == EnergyConservingScheme
-validate_divergence_and_vertical_scheme(divergence_scheme, vertical_scheme)          = (divergence_scheme, vertical_scheme)
-validate_divergence_and_vertical_scheme(::Nothing, vertical_scheme)                  = (nothing, EnergyConservingScheme())
-validate_divergence_and_vertical_scheme(::Nothing, ::EnergyConservingScheme)         = (nothing, EnergyConservingScheme())
-validate_divergence_and_vertical_scheme(divergence_scheme, ::EnergyConservingScheme) = (nothing, EnergyConservingScheme())
+              " $(a.vorticity_scheme isa WENO ? "├" : "└")── $(summary(a.vorticity_scheme))",
+              " $(a.vorticity_scheme isa WENO ? "\n └── smoothness ζ: $(a.vorticity_stencil)\n" : "\n")",
+              " Vertical advection / Divergence flux scheme: ", "\n",
+              " $(a.vertical_scheme isa WENO ? "├" : "└")── $(summary(a.vertical_scheme))",
+              "$(a.vertical_scheme isa AbstractUpwindBiasedAdvectionScheme ? 
+              "\n └── upwinding treatment: $(a.upwinding)" : "")")
 
 # Since vorticity itself requires one halo, if we use an upwinding scheme (N > 1) we require one additional
 # halo for vector invariant advection
 required_halo_size(scheme::VectorInvariant{N}) where N = N == 1 ? N : N + 1
 
-Adapt.adapt_structure(to, scheme::VectorInvariant{N, FT}) where {N, FT} =
-        VectorInvariant{N, FT}(Adapt.adapt(to, scheme.vorticity_scheme), 
-                               Adapt.adapt(to, scheme.divergence_scheme), 
-                               Adapt.adapt(to, scheme.vorticity_stencil), 
-                               Adapt.adapt(to, scheme.divergence_stencil), 
-                               Adapt.adapt(to, scheme.vertical_scheme))
-
-@inline vertical_scheme(scheme::VectorInvariant) = string(nameof(typeof(scheme.vertical_scheme)))
-
-const VectorInvariantEnergyConserving    = VectorInvariant{<:Any, <:Any, <:EnergyConservingScheme}
-const VectorInvariantEnstrophyConserving = VectorInvariant{<:Any, <:Any, <:EnstrophyConservingScheme}
-
-const VectorInvariantVerticallyEnergyConserving  = VectorInvariant{<:Any, <:Any, <:Any, Nothing, <:Any, <:Any, <:EnergyConservingScheme}
+Adapt.adapt_structure(to, scheme::VectorInvariant{N, FT, Z, ZS, V, D, M}) where {N, FT, Z, ZS, V, D, M} =
+        VectorInvariant{N, FT, M}(Adapt.adapt(to, scheme.vorticity_scheme), 
+                                  Adapt.adapt(to, scheme.vorticity_stencil), 
+                                  Adapt.adapt(to, scheme.vertical_scheme),
+                                  Adapt.adapt(to, scheme.upwinding))
 
 @inline U_dot_∇u(i, j, k, grid, scheme::VectorInvariant, U) = (
     + horizontal_advection_U(i, j, k, grid, scheme, U.u, U.v)
-    + vertical_advection_U(i, j, k, grid, scheme, U.w, U.u)
+    + vertical_advection_U(i, j, k, grid, scheme, U.w, U.u, U.v)
```

Comment:
why vertical advection of U needs V?

Reply:
it might be upwinding of the divergence, so it might need `v`